### DISCUSSION TITLE
How to get all sidebar items programmatically?

### DISCUSSION THREAD

**Discussion by gouku** [facebook/docusaurus#5414] [Q&A]
Hi, I'd like to get the all sidebar items programmatically and render it into the content of a page.

My use case is to print the table of contents (ToC) for the whole site into PDF.

But the current sidebar doesn't work because not all sidebar items are expanded. Some of our docs in sidebar are collapsed by default. Only when a user click on the top-level category item then Docusaurus can render items with that category. This prevent us from generate the whole ToC programmatically.

Also items in sidebars are hidden by default with print styles. So I'd like to know if there's an easy method to get all sidebar items then render them in to a specific statically.

Thanks!

**Comment by slorber** [comment]
It's hard for me to understand what you are trying to do exactly.

Creating a mockup with https://excalidraw.com/ can help.

Note we made sidebars "lazy" on the server so sidebars are not fully server-rendered for perf reasons (https://github.com/facebook/docusaurus/pull/5136)

Eventually you can try to swizzle this comp and remove the lazy collapsible option to get it rendered statically:

![image](https://user-images.githubusercontent.com/749374/130619628-c240ffbd-60a6-49a0-a959-80a9925165b3.png)

If this make sense and you can achieve what you want, we could add an official option to disable this lazy sidebar item behavior for certain use-cases when you really need the whole sidebars to be server-rendererd.

For print styles, you can provide your own CSS to to make them visible, and we shouldn't need to do anything about it.

**Reply by gouku** [reply]
Thanks for replying. @Siemienik seems get the point but currently it still does not work for me.

I want to display the main sidebar on the left directly inside the markdown document area rather than the table of content of current page. So that I can create a specific page contains the sidebar links as page content which can be treated as table of contents when generate PDF from it. 

The following may do the job:

```js
import TOCInline from '@theme/TOCInline';

<TOCInline toc={sidebar} />;
```

But it doesn't work. only `toc` variable is avaiable:

```js
import TOCInline from '@theme/TOCInline';

<TOCInline toc={toc} />;
```

So I'd like to know if there any other way to achieve this?

Thanks!

**Reply by slorber** [reply]
Still not sure to understand but:

- we could technically make the sidebar of a doc available in its MDX scope, allowing you to render it the way you want. 
- keep in mind that the sidebar is not a TOC (not the same json structure at all) and can't be rendered using `TOCInline`: you'll likely have to implement your own component to render it. 
- The default sidebar theme comp may also not be a good fit because it's tightly coupled to the theme layout. 
- You may be able to reuse the "SidebarItem" component

**Comment by Siemienik** [comment]
Did you attempt to use the inline table of content? Does it resolve your needs? 
https://docusaurus.io/docs/markdown-features/inline-toc

Additionally, for PDF generating, there is an open-source tool:
https://github.com/KohheePeace/docusaurus-pdf .

**Reply by gouku** [reply]
> Did you attempt to use the inline table of content

I want to display the main sidebar on the left directly inside the markdown document area rather than the table of content of current page. 

I want:

```js
import TOCInline from '@theme/TOCInline';

<TOCInline toc={sidebar} />;
```

But it doesn't work. only `toc` variable is avaiable:

```js
import TOCInline from '@theme/TOCInline';

<TOCInline toc={toc} />;
```
